In [4]:
# Read the package

# Read the data

In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [3]:
file_path="C:\\Users\\sony\\ DATA SCIENCE  NARESH IT\\Visadataset.csv"
visa_df=pd.read_csv(file_path)
visa_df

case_id continent education_of_employee has_job_experience  \
0         EZYV01      Asia           High School                  N   
1         EZYV02      Asia              Master's                  Y   
2         EZYV03      Asia            Bachelor's                  N   
3         EZYV04      Asia            Bachelor's                  N   
4         EZYV05    Africa              Master's                  Y   
...          ...       ...                   ...                ...   
25475  EZYV25476      Asia            Bachelor's                  Y   
25476  EZYV25477      Asia           High School                  Y   
25477  EZYV25478      Asia              Master's                  Y   
25478  EZYV25479      Asia              Master's                  Y   
25479  EZYV25480      Asia            Bachelor's                  Y   

      requires_job_training  no_of_employees  yr_of_estab  \
0                         N            14513         2007   
1                         N             2412         2002   
2                         Y            44444         2008   
3                         N               98         1897   
4                         N             1082         2005   
...                     ...              ...          ...   
25475                     Y             2601         2008   
25476                     N             3274         2006   
25477                     N             1121         1910   
25478                     Y             1918         1887   
25479                     N             3195         1960   

      region_of_employment  prevailing_wage unit_of_wage full_time_position  \
0                     West         592.2029         Hour                  Y   
1                Northeast       83425.6500         Year                  Y   
2                     West      122996.8600         Year                  Y   
3                     West       83434.0300         Year                  Y   
4                    South      149907.3900         Year                  Y   
...                    ...              ...          ...                ...   
25475                South       77092.5700         Year                  Y   
25476            Northeast      279174.7900         Year                  Y   
25477                South      146298.8500         Year                  N   
25478                 West       86154.7700         Year                  Y   
25479              Midwest       70876.9100         Year                  Y   

      case_status  
0          Denied  
1       Certified  
2          Denied  
3          Denied  
4       Certified  
...           ...  
25475   Certified  
25476   Certified  
25477   Certified  
25478   Certified  
25479   Certified  

[25480 rows x 12 columns]

In [4]:
visa_df.columns

Index(['case_id', 'continent', 'education_of_employee', 'has_job_experience',
       'requires_job_training', 'no_of_employees', 'yr_of_estab',
       'region_of_employment', 'prevailing_wage', 'unit_of_wage',
       'full_time_position', 'case_status'],
      dtype='object')

In [5]:
visa_df['prevailing_wage'] #as a series

0           592.2029
1         83425.6500
2        122996.8600
3         83434.0300
4        149907.3900
            ...     
25475     77092.5700
25476    279174.7900
25477    146298.8500
25478     86154.7700
25479     70876.9100
Name: prevailing_wage, Length: 25480, dtype: float64

In [6]:
visa_df['prevailing_wage'].values #as a array

array([   592.2029,  83425.65  , 122996.86  , ..., 146298.85  ,
        86154.77  ,  70876.91  ])

- count

- min

- max

- mean

- median

- std


## Using-dict

In [7]:
dict1={}
wage_count=round(visa_df['prevailing_wage'].count(),2)
wage_min=round(visa_df['prevailing_wage'].min(),2)
wage_max=round(visa_df['prevailing_wage'].min(),2)
wage_mean=round(visa_df['prevailing_wage'].mean(),2)
wage_median=round(visa_df['prevailing_wage'].median(),2)
wage_std=round(visa_df['prevailing_wage'].std(),2)

dict1['count']=wage_count
dict1['min']=wage_min
dict1['max']=wage_max
dict1['mean']=wage_mean
dict1['median']=wage_median
dict1['std']=wage_std
dict1
pd.DataFrame(dict1,index=['prevailing_wage'])


count   min   max      mean    median       std
prevailing_wage  25480  2.14  2.14  74455.81  70308.21  52815.94

In [8]:
dict1={}
wage_count=round(visa_df['prevailing_wage'].count(),2)
wage_min=round(visa_df['prevailing_wage'].min(),2)
wage_max=round(visa_df['prevailing_wage'].min(),2)
wage_mean=round(visa_df['prevailing_wage'].mean(),2)
wage_median=round(visa_df['prevailing_wage'].median(),2)
wage_std=round(visa_df['prevailing_wage'].std(),2)

list1=[wage_count,wage_min,wage_max,wage_mean,wage_median,wage_std]
dict1['prevailing_wage']=list1
dict1
pd.DataFrame(dict1)


prevailing_wage
0         25480.00
1             2.14
2             2.14
3         74455.81
4         70308.21
5         52815.94

## Using_list

In [9]:
wage_count=round(visa_df['prevailing_wage'].count(),2)
wage_min=round(visa_df['prevailing_wage'].min(),2)
wage_max=round(visa_df['prevailing_wage'].min(),2)
wage_mean=round(visa_df['prevailing_wage'].mean(),2)
wage_median=round(visa_df['prevailing_wage'].median(),2)
wage_std=round(visa_df['prevailing_wage'].std(),2)

list1=[wage_count,wage_min,wage_max,wage_mean,wage_median,wage_std]
pd.DataFrame(list1,columns=['prevailing_wage'],
            index=['count','min','max','mean','median','std'])

prevailing_wage
count          25480.00
min                2.14
max                2.14
mean           74455.81
median         70308.21
std            52815.94

In [10]:
# step-1: numerical column list

dtypes=dict(visa_df.dtypes)
num=[i for i in dtypes if dtypes[i]!='O']
num

['no_of_employees', 'yr_of_estab', 'prevailing_wage']

In [11]:
# secoand way

num_lst=[i for i in dtypes if dtypes[i] in ['int64','int32','float64']]
num_lst

['no_of_employees', 'yr_of_estab', 'prevailing_wage']

In [12]:
dtypes=dict(visa_df.dtypes)
num=[i for i in dtypes if dtypes[i]!='O']
print(num)


for i in num:
    count=round(visa_df[i].count(),2)
    Min=round(visa_df[i].min(),2)
    Max=round(visa_df[i].min(),2)
    mean=round(visa_df[i].mean(),2)
    median=round(visa_df[i].median(),2)
    std=round(visa_df[i].std(),2)

    list1=[count,Min,Max,mean,median,std]
    df=pd.DataFrame(list1,columns=[i],
                index=['count','min','max','mean','median','std'])
    print(df)

['no_of_employees', 'yr_of_estab', 'prevailing_wage']
        no_of_employees
count          25480.00
min              -26.00
max              -26.00
mean            5667.04
median          2109.00
std            22877.93
        yr_of_estab
count      25480.00
min         1800.00
max         1800.00
mean        1979.41
median      1997.00
std           42.37
        prevailing_wage
count          25480.00
min                2.14
max                2.14
mean           74455.81
median         70308.21
std            52815.94


In [13]:
dtypes=dict(visa_df.dtypes)
num=[i for i in dtypes if dtypes[i]!='O']
print(num)

dict1={}
for i in num:
    count=round(visa_df[i].count(),2)
    Min=round(visa_df[i].min(),2)
    Max=round(visa_df[i].min(),2)
    mean=round(visa_df[i].mean(),2)
    median=round(visa_df[i].median(),2)
    std=round(visa_df[i].std(),2)

    list1=[count,Min,Max,mean,median,std]
    dict1[i]=list1
    
    df=pd.DataFrame(dict1,
                index=['count','min','max','mean','median','std'])
print(df)

['no_of_employees', 'yr_of_estab', 'prevailing_wage']
        no_of_employees  yr_of_estab  prevailing_wage
count          25480.00     25480.00         25480.00
min              -26.00      1800.00             2.14
max              -26.00      1800.00             2.14
mean            5667.04      1979.41         74455.81
median          2109.00      1997.00         70308.21
std            22877.93        42.37         52815.94


In [14]:
dtypes=dict(visa_df.dtypes)
num=[i for i in dtypes if dtypes[i]!='O']
print(num)

dict1={}
for i in num:
    count=round(visa_df[i].count(),2)
    Min=round(visa_df[i].min(),2)
    Max=round(visa_df[i].min(),2)
    mean=round(visa_df[i].mean(),2)
    median=round(visa_df[i].median(),2)
    std=round(visa_df[i].std(),2)

    list1=[count,Min,Max,mean,median,std]
    dict1[i]=list1
    
    df=pd.DataFrame(dict1,
                index=['count','min','max','mean','median','std'])
dict1

['no_of_employees', 'yr_of_estab', 'prevailing_wage']


{'no_of_employees': [25480, -26, -26, 5667.04, 2109.0, 22877.93],
 'yr_of_estab': [25480, 1800, 1800, 1979.41, 1997.0, 42.37],
 'prevailing_wage': [25480, 2.14, 2.14, 74455.81, 70308.21, 52815.94]}

In [15]:
df

no_of_employees  yr_of_estab  prevailing_wage
count          25480.00     25480.00         25480.00
min              -26.00      1800.00             2.14
max              -26.00      1800.00             2.14
mean            5667.04      1979.41         74455.81
median          2109.00      1997.00         70308.21
std            22877.93        42.37         52815.94

In [16]:
visa_df.describe()

no_of_employees   yr_of_estab  prevailing_wage
count     25480.000000  25480.000000     25480.000000
mean       5667.043210   1979.409929     74455.814592
std       22877.928848     42.366929     52815.942327
min         -26.000000   1800.000000         2.136700
25%        1022.000000   1976.000000     34015.480000
50%        2109.000000   1997.000000     70308.210000
75%        3504.000000   2005.000000    107735.512500
max      602069.000000   2016.000000    319210.270000

In [17]:
visa_df['prevailing_wage'].mean()

# Reading a specific column
# we have a mean method

74455.81459209221

In [18]:
# np.mean(<specific column data>)

In [19]:
np.mean(visa_df['prevailing_wage'])
np.median(visa_df['prevailing_wage'])
np.std(visa_df['prevailing_wage'])
np.min(visa_df['prevailing_wage'])
np.max(visa_df['prevailing_wage'])

319210.27

$percentile$-$quantile$

In [ ]:
- percentile ranges from 1 to 100

- quantile q1=25P  q2=50P  q3=75p

- np.percentile(<direct number between 1 to 100>,data)

- ex: np.percentile(75,data)

- np.quantile(<decimal number>,data)
    
- np.quantile(0.75,data)    

In [20]:
q1=round(np.percentile(visa_df['prevailing_wage'],25),2)
q2=round(np.percentile(visa_df['prevailing_wage'],50),2)
q3=round(np.percentile(visa_df['prevailing_wage'],75),2)
print(q1,q2,q3)

34015.48 70308.21 107735.51


In [21]:
q1=round(np.quantile(visa_df['prevailing_wage'],0.25),2)
q2=round(np.quantile(visa_df['prevailing_wage'],0.50),2)
q3=round(np.quantile(visa_df['prevailing_wage'],0.75),2)
print(q1,q2,q3)

34015.48 70308.21 107735.51


In [22]:
dtypes=dict(visa_df.dtypes)
num=[i for i in dtypes if dtypes[i]!='O']
print(num)

dict1={}
for i in num:
    count=round(visa_df[i].count(),2)
    Min=round(visa_df[i].min(),2)
    Max=round(visa_df[i].min(),2)
    mean=round(visa_df[i].mean(),2)
    median=round(visa_df[i].median(),2)
    std=round(visa_df[i].std(),2)
    #####################################
    q1=round(np.percentile(visa_df[i],25),2)
    q2=round(np.percentile(visa_df[i],50),2)
    q3=round(np.percentile(visa_df[i],75),2)
     

    list1=[count,Min,Max,mean,median,std,q1,q2,q3]
    dict1[i]=list1
    
    df=pd.DataFrame(dict1,
                index=['count','min','max','mean','median','std','25%','50%','75%'])


['no_of_employees', 'yr_of_estab', 'prevailing_wage']


In [23]:
df

no_of_employees  yr_of_estab  prevailing_wage
count          25480.00     25480.00         25480.00
min              -26.00      1800.00             2.14
max              -26.00      1800.00             2.14
mean            5667.04      1979.41         74455.81
median          2109.00      1997.00         70308.21
std            22877.93        42.37         52815.94
25%             1022.00      1976.00         34015.48
50%             2109.00      1997.00         70308.21
75%             3504.00      2005.00        107735.51

In [24]:
q1=round(np.percentile(visa_df['prevailing_wage'],25),2)
q1

34015.48

## what is the meaning of this

In [26]:
# 25 percentage of observation from the total data have a value below 34015

In [28]:
# total_observations=25480
# 25percentage(25480)
# 25*25480/100=6370

# 6370 people have wages less than 34015

In [29]:
len(visa_df[visa_df['prevailing_wage']<34015])
    
# how many True= 670    

6370

### Empiricle-rule

<img src="https://builtin.com/sites/www.builtin.com/files/styles/ckeditor_optimize/public/inline-images/1_empirical-rule.jpg" jsaction="" class="sFlh5c FyHeAf iPVvYb" style="max-width: 810px; height: 148px; margin: 0px; width: 296px;" alt="Empirical Rule (68-95-99.7) Explained | Built In" jsname="kn3ccd" aria-hidden="false">

- search empiricale-rule in google

- Right click select insepect

- click on inspect

- right side you will omg src

- right click on omg src and select edit as html

- dont move your curosr

- CTRL+A

- CTRL+c

- CTRL+V

- ESC+M

- SHIFT+ENTER

### when data follows a normal distribution

- u-1*sigma to u+1*sigma : 68%
    
- u-1*sigma to u+1*sigma : 95%
    
- u-1*sigma to u+1*sigma : 99.7%    

In [30]:
wage_mean,wage_std

(74455.81, 52815.94)

In [31]:
####################### 68% #######################
val_minus_1=round(wage_mean-1*wage_std,2)
val_plus_1=round(wage_mean+1*wage_std,2)

######################## 95% ######################

val_minus_2=round(wage_mean-2*wage_std,2)
val_plus_2=round(wage_mean+2*wage_std,2)

######################## 99.7% ######################
val_minus_3=round(wage_mean-3*wage_std,2)
val_plus_3=round(wage_mean+3*wage_std,2)

print(val_minus_1,val_plus_1,val_minus_2,val_plus_2,val_minus_3,val_plus_3)

21639.87 127271.75 -31176.07 180087.69 -83992.01 232903.63


- 68 ppercentage of observations have value between [21639.87,127271.75]

- 95 ppercentage of observations have value between [-31176.07,180087.69]

- 99.7 ppercentage of observations have value between [-83992.01,232903.63]

In [32]:
68*25480/100

17326.4

In [33]:
con1=visa_df['prevailing_wage']>val_minus_1
con2=visa_df['prevailing_wage']<val_plus_1
len(visa_df[con1&con2])
len(visa_df[con1&con2])/len(visa_df)

0.673901098901099

In [34]:
con1=visa_df['prevailing_wage']>val_minus_2
con2=visa_df['prevailing_wage']<val_plus_2
len(visa_df[con1&con2])
len(visa_df[con1&con2])/len(visa_df)

0.9647566718995291

In [35]:
con1=visa_df['prevailing_wage']>val_minus_3
con2=visa_df['prevailing_wage']<val_plus_3
len(visa_df[con1&con2])
len(visa_df[con1&con2])/len(visa_df)

0.9884615384615385

In [ ]:
68-95-99.7
67-96-98

- we read p_wage column

- we perrorm statistical analysis

- we perform empirical rule analysis

- p_wage  67-96-98

- wage_mean=74455

- wage_median=70308

- meadian<mean

- data might be right skewed

- but percentage of data almost valid with empirical rule

- it is looks like normal distribution and slightly right skewed

- in order to see that draw that histogram